File: HuanLuyen/Classes/BDTC/CThumbNails.cs
using DBiGraphicObjs.DBiGraphicObjects;
using System;
using System.Collections;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Runtime.CompilerServices;
using System.Windows.Forms;
namespace HuanLuyen
{
    public class CThumbNails : CollectionBase
    {
        public delegate void SelectedIndexChangedEventHandler(MouseEventArgs e);
        public delegate void ChonEventHandler(int index);
        private CThumbNails.SelectedIndexChangedEventHandler SelectedIndexChangedEvent;
        private CThumbNails.ChonEventHandler ChonEvent;
        [AccessedThroughProperty("m_PicBox")]
        private PictureBox _m_PicBox;
        private int currentTopLeftItem = 0;
        private int m_SelectedIndex = 0;
        private string m_GenericImage;
        private int x = 0;
        private int y = 0;
        private int h = 0;
        private int w = 0;
        private int rowsPerPage = 0;
        private int colsPerPage = 0;
        public event CThumbNails.SelectedIndexChangedEventHandler SelectedIndexChanged
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            add
            {
                this.SelectedIndexChangedEvent = (CThumbNails.SelectedIndexChangedEventHandler)Delegate.Combine(this.SelectedIndexChangedEvent, value);
            }
            [MethodImpl(MethodImplOptions.Synchronized)]
            remove
            {
                this.SelectedIndexChangedEvent = (CThumbNails.SelectedIndexChangedEventHandler)Delegate.Remove(this.SelectedIndexChangedEvent, value);
            }
        }
        public event CThumbNails.ChonEventHandler Chon
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            add
            {
                this.ChonEvent = (CThumbNails.ChonEventHandler)Delegate.Combine(this.ChonEvent, value);
            }
            [MethodImpl(MethodImplOptions.Synchronized)]
            remove
            {
                this.ChonEvent = (CThumbNails.ChonEventHandler)Delegate.Remove(this.ChonEvent, value);
            }
        }
        protected virtual PictureBox m_PicBox
        {
            get
            {
                return this._m_PicBox;
            }
            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                PaintEventHandler value2 = new PaintEventHandler(this.m_PicBox_Paint);
                EventHandler value3 = new EventHandler(this.m_PicBox_DoubleClick);
                MouseEventHandler value4 = new MouseEventHandler(this.m_PicBox_MouseUp);
                EventHandler value5 = new EventHandler(this.m_PicBox_Resize);
                if (this._m_PicBox != null)
                {
                    this._m_PicBox.Paint -= value2;
                    this._m_PicBox.DoubleClick -= value3;
                    this._m_PicBox.MouseUp -= value4;
                    this._m_PicBox.Resize -= value5;
                }
                this._m_PicBox = value;
                if (this._m_PicBox != null)
                {
                    this._m_PicBox.Paint += value2;
                    this._m_PicBox.DoubleClick += value3;
                    this._m_PicBox.MouseUp += value4;
                    this._m_PicBox.Resize += value5;
                }
            }
        }
        public int HorizontalSpacing
        {
            get
            {
                return this.x;
            }
            set
            {
                this.x = value;
                this.CalculateRowsAndColumns();
            }
        }
        public int VerticalSpacing
        {
            get
            {
                return this.y;
            }
            set
            {
                this.y = value;
                this.CalculateRowsAndColumns();
            }
        }
        public int ImageHeight
        {
            get
            {
                return this.h;
            }
            set
            {
                this.h = value;
                this.CalculateRowsAndColumns();
            }
        }
        public int ImageWidth
        {
            get
            {
                return this.w;
            }
            set
            {
                this.w = value;
                this.CalculateRowsAndColumns();
            }
        }
        public int SelectedIndex
        {
            get
            {
                return this.m_SelectedIndex;
            }
            set
            {
                IList list = this.List;
                checked
                {
                    if (list != null)
                    {
                        int num = this.currentTopLeftItem;
                        int selectedIndex = this.m_SelectedIndex;
                        if (value < 0)
                        {
                            if (selectedIndex != value)
                            {
                                this.m_SelectedIndex = value;
                                this.m_PicBox.Invalidate();
                            }
                        }
                        else if (value >= 0 & value < list.Count)
                        {
                            if (value - this.currentTopLeftItem >= this.rowsPerPage * this.colsPerPage)
                            {
                                do
                                {
                                    this.currentTopLeftItem += this.rowsPerPage * this.colsPerPage;
                                }
                                while (value - this.currentTopLeftItem >= this.rowsPerPage * this.colsPerPage);
                            }
                            else if (value - this.currentTopLeftItem < 0)
                            {
                                do
                                {
                                    this.currentTopLeftItem -= this.rowsPerPage * this.colsPerPage;
                                }
                                while (value - this.currentTopLeftItem < 0);
                            }
                            if (this.currentTopLeftItem < 0)
                            {
                                this.currentTopLeftItem = 0;
                            }
                            if (selectedIndex != value)
                            {
                                this.m_SelectedIndex = value;
                                if (num != this.currentTopLeftItem)
                                {
                                    this.m_PicBox.Invalidate();
                                }
                                else
                                {
                                    this.RefreshItem(selectedIndex);
                                    this.RefreshItem(value);
                                }
                            }
                        }
                    }
                }
            }
        }
        public CThumbNail this[int index]
        {
            get
            {
                return (CThumbNail)this.List[index];
            }
        }
        public void Add(CThumbNail aThumbNail)
        {
            this.List.Add(aThumbNail);
        }
        public int GetCount()
        {
            return this.List.Count;
        }
        public void InsertAt(int index, CThumbNail aThumbNail)
        {
            if (index > this.Count | index < 0)
            {
                MessageBox.Show("Index not valid!");
            }
            else if (index == this.Count)
            {
                this.List.Add(aThumbNail);
            }
            else
            {
                this.List.Insert(index, aThumbNail);
            }
        }
        public void Remove(int index)
        {
            if (index > checked(this.Count - 1) | index < 0)
            {
                MessageBox.Show("Index not valid!");
            }
            else
            {
                this.List.RemoveAt(index);
            }
        }
        public void Remove(CThumbNail aThumbNail)
        {
            this.List.Remove(aThumbNail);
        }
        public CThumbNail GetItemByID(int pId)
        {
            foreach (CThumbNail cThumbNail in this.List)
            {
                if (cThumbNail.ID == pId)
                {
                    return cThumbNail;
                }
            }

            return null;
        }
        public int MoveItem(int pItemId, int pDestItemID)
        {
            int num = -1;
            CThumbNail itemByID = this.GetItemByID(pItemId);
            this.Remove(itemByID);
            checked
            {
                foreach (CThumbNail cThumbNail in this.List)
                {
                    num++;
                    if (cThumbNail.ID == pDestItemID)
                    {
                        this.InsertAt(num, itemByID);
                        return num;
                    }
                }

                return -1;
            }
        }
        public int GetMaxID()
        {
            int num = 0;
            foreach (CThumbNail cThumbNail in this.List)
            {
                if (cThumbNail.ID > num)
                {
                    num = cThumbNail.ID;
                }
            }

            return num;
        }
        private void m_PicBox_Paint(object sender, PaintEventArgs e)
        {
            Graphics graphics = e.Graphics;
            graphics.SmoothingMode = SmoothingMode.AntiAlias;
            IList list = this.List;
            checked
            {
                if (list != null && list.Count > 0)
                {
                    int count = list.Count;
                    int num = this.currentTopLeftItem;
                    int arg_41_0 = 0;
                    int num2 = this.rowsPerPage - 1;
                    for (int i = arg_41_0; i <= num2; i++)
                    {
                        int arg_4F_0 = 0;
                        int num3 = this.colsPerPage - 1;
                        for (int j = arg_4F_0; j <= num3; j++)
                        {
                            if (num < count)
                            {
                                this.DrawOneItem(num, i, j, graphics);
                                num++;
                            }
                        }
                    }
                    Font font = new Font("Webdings", 20f, FontStyle.Regular, GraphicsUnit.Pixel);
                    SolidBrush solidBrush = new SolidBrush(this.m_PicBox.ForeColor);
                    if (num <= count - 1)
                    {
                        graphics.DrawString("6", font, solidBrush, 0f, (float)(this.m_PicBox.Height - 24));
                    }
                    if (this.currentTopLeftItem > 0)
                    {
                        graphics.DrawString("5", font, solidBrush, 0f, 0f);
                    }
                    font.Dispose();
                    solidBrush.Dispose();
                }
            }
        }
        private void DrawOneItem(int index, int row, int col, Graphics gr)
        {
            Font font = this.m_PicBox.Font;
            SolidBrush solidBrush = new SolidBrush(this.m_PicBox.ForeColor);
            StringFormat stringFormat = new StringFormat();
            stringFormat.Alignment = StringAlignment.Center;
            stringFormat.FormatFlags = StringFormatFlags.LineLimit;
            CThumbNail cThumbNail = (CThumbNail)this.List[index];
            string symbols = cThumbNail.Symbols;
            CGraphicObjs cGraphicObjs = CGraphicObjs.Str2Objects(symbols, 0, 0, modHuanLuyen.cDecSepa, modHuanLuyen.cGrpSepa);
            Size size = cGraphicObjs.GetSize();
            checked
            {
                int num = 0;
                int num2 = 0;
                if (size.Height > this.h)
                {
                    num = this.h;
                    num2 = (int)Math.Round(unchecked((double)this.h / (double)size.Height * (double)size.Width));
                }
                else
                {
                    num = size.Height;
                    num2 = size.Width;
                }
                if (num2 > this.w)
                {
                    num = (int)Math.Round(unchecked((double)this.w / (double)num2 * (double)num));
                    num2 = this.w;
                }
                Rectangle rect = new Rectangle(2 * this.x + col * (2 * this.x + this.w), 1 * this.y + row * (2 * this.y + this.h), this.w, this.h);
                gr.FillRectangle(new SolidBrush(Color.LightGray), rect);
                Rectangle imageRect = new Rectangle(2 * this.x + col * (2 * this.x + this.w) + (this.w - num2) / 2, 1 * this.y + row * (2 * this.y + this.h) + (this.h - num) / 2, num2, num);
                this.DrawThumbNail(gr, cGraphicObjs, imageRect);
                Pen pen = null;
                if (index == this.m_SelectedIndex)
                {
                    pen = new Pen(Color.Yellow);
                    pen.Width = 4f;
                    gr.DrawRectangle(pen, rect);
                }
                if (pen != null)
                {
                    pen.Dispose();
                }
                int num3 = this.y * 2;
                string arg_24A_1 = cThumbNail.Value;
                Font arg_24A_2 = font;
                Brush arg_24A_3 = solidBrush;
                RectangleF layoutRectangle = new RectangleF((float)(this.x + col * (2 * this.x + this.w)), (float)(2 + 1 * this.y + this.h + row * (2 * this.y + this.h)), (float)(this.w + 2 * this.x), (float)num3);
                gr.DrawString(arg_24A_1, arg_24A_2, arg_24A_3, layoutRectangle, stringFormat);
            }
        }
        private void DrawThumbNail(Graphics g, CGraphicObjs mObjs, Rectangle imageRect)
        {
            float num = (float)mObjs.GetSize().Width;
            float num2 = 0;
            if (num > 0f)
            {
                num2 = (float)imageRect.Width / num;
            }
            else
            {
                num2 = 1f;
            }
            GraphicsContainer container = g.BeginContainer();
            g.TranslateTransform((float)imageRect.X, (float)imageRect.Y);
            g.ScaleTransform(num2, num2);
            foreach (GraphicObject graphicObject in mObjs)
            {
                if (graphicObject.GetObjType() != OBJECTTYPE.Text)
                {
                    graphicObject.Draw(g);
                }
            }

            g.EndContainer(container);
        }
        public void CalculateRowsAndColumns()
        {
            checked
            {
                int num = this.m_PicBox.Height - 24;
                int num2 = (num - this.y) / (2 * this.y + this.h);
                int num3 = this.m_PicBox.Width / (2 * this.x + this.w);
                if (num2 != this.rowsPerPage || num3 != this.colsPerPage)
                {
                    this.rowsPerPage = num2;
                    this.colsPerPage = num3;
                }
            }
        }
        private Rectangle GetItemRect(int index)
        {
            checked
            {
                int num = (int)Math.Round(Math.Floor((double)(index - this.currentTopLeftItem) / (double)this.colsPerPage));
                int num2 = (index - this.currentTopLeftItem) % this.colsPerPage;
                Rectangle result = new Rectangle(this.x + num2 * (2 * this.x + this.w), 1 * this.y + num * (2 * this.y + this.h), this.w + 2 * this.x, this.h + this.y * 2);
                return result;
            }
        }
        private void m_PicBox_Resize(object sender, EventArgs e)
        {
            this.CalculateRowsAndColumns();
            this.m_PicBox.Invalidate();
        }
        public CThumbNails(PictureBox pPicBox)
        {
            this.currentTopLeftItem = 0;
            this.m_SelectedIndex = 0;
            this.m_GenericImage = "";
            this.x = 10;
            this.y = 16;
            this.h = 40;
            this.w = 40;
            this.rowsPerPage = 0;
            this.colsPerPage = 0;
            this.m_PicBox = pPicBox;
            this.CalculateRowsAndColumns();
        }
        private int HitTest(Point loc)
        {
            bool flag = false;
            int num = 0;
            checked
            {
                while (num < this.List.Count & !flag)
                {
                    if (this.GetItemRect(num).Contains(loc))
                    {
                        flag = true;
                    }
                    else
                    {
                        num++;
                    }
                }
                if (flag)
                {
                    return num;
                }
                return -1;
            }
        }
        protected void RefreshItem(int index)
        {
            Rectangle itemRect = this.GetItemRect(index);
            itemRect.Inflate(5, 5);
            this.m_PicBox.Invalidate(itemRect);
        }
        private void m_PicBox_MouseUp(object sender, MouseEventArgs e)
        {
            Point loc = new Point(e.X, e.Y);
            int num = this.HitTest(loc);
            if (num > -1)
            {
                this.SelectedIndex = num;
            }
            else
            {
                this.SelectedIndex = -1;
            }
            CThumbNails.SelectedIndexChangedEventHandler selectedIndexChangedEvent = this.SelectedIndexChangedEvent;
            if (selectedIndexChangedEvent != null)
            {
                selectedIndexChangedEvent(e);
            }
        }
        private void m_PicBox_DoubleClick(object sender, EventArgs e)
        {
            if (this.m_SelectedIndex > -1)
            {
                CThumbNails.ChonEventHandler chonEvent = this.ChonEvent;
                if (chonEvent != null)
                {
                    chonEvent(this.m_SelectedIndex);
                }
            }
        }
    }
}

File: BDTCLib/CThumbNails.vb
'Option Strict On
Option Explicit On

Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D

Imports DBiGraphicObjs.DBiGraphicObjects
Imports System.Windows.Forms

Public Class CThumbNail
    Private mValue As String
    Private mID As Integer
    Private mSymbols As String
    Private mVanToc As Integer = 4

    Public Sub New(ByVal strValue As String, ByVal intID As Integer, ByVal strSymbols As String)
        mValue = strValue
        mID = intID
        mSymbols = strSymbols
    End Sub

    Public Sub New(ByVal strValue As String, ByVal intID As Integer, ByVal strSymbols As String, ByVal intVanToc As Integer)
        mValue = strValue
        mID = intID
        mSymbols = strSymbols
        mVanToc = intVanToc
    End Sub

    Public Sub New()
        mValue = ""
        mID = 0
        mSymbols = ""
    End Sub

    Property ID() As Integer
        Get
            Return mID
        End Get
        Set(ByVal Value As Integer)
            mID = Value
        End Set
    End Property

    Property Value() As String
        Get
            Return mValue
        End Get
        Set(ByVal Value As String)
            mValue = Value
        End Set
    End Property

    Property Symbols() As String
        Get
            Return mSymbols
        End Get
        Set(ByVal Value As String)
            mSymbols = Value
        End Set
    End Property

    Property VanToc() As Integer
        Get
            Return mVanToc
        End Get
        Set(ByVal Value As Integer)
            mVanToc = Value
        End Set
    End Property

    Public Overrides Function ToString() As String
        Return mValue
    End Function

End Class

Public Class CThumbNails
    Inherits System.Collections.CollectionBase

    Public Event SelectedIndexChanged(ByVal e As System.EventArgs)

    Private WithEvents m_PicBox As PictureBox

    Private currentTopLeftItem As Integer = 0

    Private m_SelectedIndex As Integer = -1
    Private m_GenericImage As String = ""

    Private x As Integer = 10
    Private y As Integer = 32 '16
    Private h As Integer = 40
    Private w As Integer = 80 '40
    Private rowsPerPage As Integer = 0
    Private colsPerPage As Integer = 0

    Public Property HorizontalSpacing() As Integer
        Get
            Return x
        End Get
        Set(ByVal Value As Integer)
            x = Value
            CalculateRowsAndColumns()
        End Set
    End Property

    Public Property VerticalSpacing() As Integer
        Get
            Return y
        End Get
        Set(ByVal Value As Integer)
            y = Value
            CalculateRowsAndColumns()
        End Set
    End Property

    Public Property ImageHeight() As Integer
        Get
            Return h
        End Get
        Set(ByVal Value As Integer)
            h = Value
            CalculateRowsAndColumns()
        End Set
    End Property

    Public Property ImageWidth() As Integer
        Get
            Return w
        End Get
        Set(ByVal Value As Integer)
            w = Value
            CalculateRowsAndColumns()
        End Set
    End Property

    Public Property SelectedIndex() As Integer
        Get
            Return m_SelectedIndex
        End Get
        Set(ByVal Value As Integer)
            'm_SelectedIndex = Value
            Dim myList As IList
            myList = List
            If Not myList Is Nothing Then
                Dim currentPosition As Integer
                Dim oldTopLeftItem, oldValue As Integer
                oldTopLeftItem = currentTopLeftItem
                oldValue = m_SelectedIndex
                If Value < 0 Then
                    m_SelectedIndex = Value
                ElseIf Value >= 0 And Value < myList.Count Then
                    'Me.DataManager.Position = Value
                    currentPosition = Value
                    If (currentPosition - currentTopLeftItem) >= (rowsPerPage * colsPerPage) Then
                        Do
                            currentTopLeftItem += (rowsPerPage * colsPerPage)
                        Loop Until (currentPosition - currentTopLeftItem) < (rowsPerPage * colsPerPage)
                    ElseIf (currentPosition - currentTopLeftItem) < 0 Then
                        Do
                            currentTopLeftItem -= (rowsPerPage * colsPerPage)
                        Loop Until (currentPosition - currentTopLeftItem) >= 0
                    End If
                    If currentTopLeftItem < 0 Then currentTopLeftItem = 0

                    If oldValue <> Value Then
                        m_SelectedIndex = Value
                        'Me.DataManager.Position = m_SelectedIndex
                        RaiseEvent SelectedIndexChanged(New System.EventArgs)
                        If oldTopLeftItem <> currentTopLeftItem Then
                            m_PicBox.Invalidate() 'redraw the whole thing
                        Else
                            Me.RefreshItem(oldValue)
                            Me.RefreshItem(Value)
                        End If
                    End If
                End If
            End If
        End Set
    End Property

    Public Sub Add(ByVal aThumbNail As CThumbNail)
        List.Add(aThumbNail)
    End Sub

    Public Function GetCount() As Integer
        Return List.Count
    End Function

    Public Sub InsertAt(ByVal index As Integer, ByVal aThumbNail As CThumbNail)
        If index > Count Or index < 0 Then
            System.Windows.Forms.MessageBox.Show("Index not valid!")
        ElseIf index = Count Then
            List.Add(aThumbNail)
        Else
            List.Insert(index, aThumbNail)
        End If
    End Sub

    Public Sub Remove(ByVal index As Integer)
        If index > Count - 1 Or index < 0 Then
            System.Windows.Forms.MessageBox.Show("Index not valid!")
        Else
            List.RemoveAt(index)
        End If
    End Sub

    Public Sub Remove(ByVal aThumbNail As CThumbNail)
        Me.List.Remove(aThumbNail)
    End Sub

    Default Public ReadOnly Property Item(ByVal index As Integer) As CThumbNail
        Get
            Return CType(List.Item(index), CThumbNail)
        End Get
    End Property

    Public Function GetItemByID(ByVal pId As Integer) As CThumbNail
        'Dim aThumbNail As CThumbNail
        For Each aThumbNail As CThumbNail In List
            If aThumbNail.ID = pId Then
                Return aThumbNail
                Exit For
            End If
        Next
        Return Nothing
    End Function

    Public Function MoveItem(ByVal pItemId As Integer, ByVal pDestItemID As Integer) As Integer
        Dim i As Integer = -1
        Dim pItem As CThumbNail = GetItemByID(pItemId)
        Remove(pItem)
        For Each aThumbNail As CThumbNail In List
            i += 1
            If aThumbNail.ID = pDestItemID Then
                InsertAt(i, pItem)
                Return i
                Exit For
            End If
        Next
        Return -1
    End Function

    Public Function GetMaxID() As Integer
        Dim iMax As Integer = 0
        For Each aThumbNail As CThumbNail In List
            If aThumbNail.ID > iMax Then
                iMax = aThumbNail.ID
            End If
        Next
        Return iMax
    End Function

    Private Sub m_PicBox_MouseWheel(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles m_PicBox.MouseWheel
        Dim oldSelectedIndex As Integer = m_SelectedIndex
        If e.Delta > 0 Then
            If oldSelectedIndex > 0 Then
                SelectedIndex = oldSelectedIndex - 1
            End If
        ElseIf e.Delta < 0 Then
            If oldSelectedIndex < (List.Count - 1) Then
                SelectedIndex = oldSelectedIndex + 1
            End If
        End If
    End Sub

    Private Sub m_PicBox_Paint(ByVal sender As Object, ByVal e As System.Windows.Forms.PaintEventArgs) Handles m_PicBox.Paint
        Dim myList As IList
        Dim gr As Graphics = e.Graphics
        gr.SmoothingMode = SmoothingMode.AntiAlias
        'gr.Clear(Color.White)
        myList = List
        If Not myList Is Nothing Then 'if there is any data
            If myList.Count > 0 Then
                Dim itemCount As Integer
                itemCount = myList.Count

                Dim itemsDisplayed As Integer 'current position in the list
                itemsDisplayed = currentTopLeftItem

                Dim i, j As Integer 'loop indexes
                'Dim height, width As Integer

                For i = 0 To rowsPerPage - 1
                    For j = 0 To colsPerPage - 1
                        If itemsDisplayed < itemCount Then
                            DrawOneItem(itemsDisplayed, i, j, gr)
                            itemsDisplayed += 1
                        End If
                    Next
                Next

                'draw page down / page up indicators
                Dim webdingsFont As New Font("Webdings", 20, FontStyle.Regular, GraphicsUnit.Pixel)
                Dim textBrush As New SolidBrush(m_PicBox.ForeColor)

                If itemsDisplayed < itemCount - 1 Then
                    'draw down arrow
                    gr.DrawString("6", webdingsFont, textBrush, 0, m_PicBox.Height - 24)
                End If

                If currentTopLeftItem > 0 Then
                    'draw down arrow
                    gr.DrawString("5", webdingsFont, textBrush, 0, 0)
                End If
                webdingsFont.Dispose()
                textBrush.Dispose()
            End If
        End If

    End Sub

    Private Sub DrawOneItem(ByVal index As Integer, ByVal row As Integer, ByVal col As Integer, ByVal gr As Graphics)
        Dim textFont As Font = m_PicBox.Font
        Dim textBrush As New SolidBrush(m_PicBox.ForeColor)
        Dim Height, Width As Integer

        Dim myStringFormat As StringFormat = New StringFormat
        myStringFormat.Alignment = StringAlignment.Center
        myStringFormat.FormatFlags = StringFormatFlags.LineLimit

        Dim aThumbNail As CThumbNail = CType(List.Item(index), CThumbNail)
        Dim strKyHieu As String = aThumbNail.Symbols
        Dim mdrawingObjects As CGraphicObjs = CGraphicObjs.Str2Objects(strKyHieu, 0, 0) 'Str2Objects(strKyHieu)
        Dim mSiZe As Size = mdrawingObjects.GetSize

        'scale image to fit into defined size
        With mSiZe
            If .Height > h Then
                Height = h
                Width = CInt((h / .Height) * .Width)
            Else
                Height = .Height
                Width = .Width
            End If
            If Width > w Then
                Height = CInt((w / Width) * Height)
                Width = w
            End If
        End With

        Dim mRect _
            As New Rectangle((2 * x) + (col * ((2 * x) + w)), _
            (1 * y) + (row * ((2 * y) + h)), _
            w, h)
        gr.FillRectangle(New SolidBrush(Color.LightGray), mRect)

        Dim imageRect _
            As New Rectangle((2 * x) + (col * ((2 * x) + w)) + ((w - Width) \ 2), _
            (1 * y) + (row * ((2 * y) + h)) + ((h - Height) \ 2), _
            Width, Height)

        DrawThumbNail(gr, mdrawingObjects, imageRect)

        Dim myNewPen As Pen
        'If index = Me.DataManager.Position Then  'selected
        If index = m_SelectedIndex Then   'selected
            myNewPen = New Pen(Color.Yellow)
            'myNewPen = New Pen(Color.Gray)
            myNewPen.Width = 4
            gr.DrawRectangle(myNewPen, mRect)
            myNewPen.Dispose()
        End If
        'End If

        Dim textHeight As Integer = y * 2
        gr.DrawString(aThumbNail.Value, _
            textFont, textBrush, _
            New RectangleF((x) + (col * ((2 * x) + w)), _
                2 + (1 * y) + h + (row * ((2 * y) + h)), _
                w + (2 * x), textHeight), myStringFormat)

        mdrawingObjects = Nothing
    End Sub

    Private Sub DrawThumbNail(ByVal g As Graphics, ByVal mObjs As CGraphicObjs, ByVal imageRect As Rectangle)
        'Dim drawObj As MapObjects.GraphicObject
        Dim Scale As Single
        Dim ImageW As Single = mObjs.GetSize.Width
        If ImageW > 0 Then
            Scale = imageRect.Width / ImageW
        Else
            Scale = 1
        End If
        'Dim Scale As Single = imageRect.Width / mObjs.GetSize.Width

        Dim gCon As Drawing2D.GraphicsContainer
        gCon = g.BeginContainer
        g.TranslateTransform(imageRect.X, imageRect.Y)
        g.ScaleTransform(Scale, Scale)
        For Each drawObj As GraphicObject In mObjs
            If Not drawObj.GetObjType = OBJECTTYPE.Text Then
                drawObj.Draw(g)
            End If
        Next
        g.EndContainer(gCon)
    End Sub

    Public Sub CalculateRowsAndColumns()
        Dim new_rowsPerPage As Integer = (m_PicBox.Height - y) \ ((2 * y) + h)
        Dim new_colsPerPage As Integer = m_PicBox.Width \ ((2 * x) + w)
        If (new_rowsPerPage <> rowsPerPage) OrElse (new_colsPerPage <> colsPerPage) Then
            rowsPerPage = new_rowsPerPage
            colsPerPage = new_colsPerPage
        End If
    End Sub

    Private Function GetItemRect(ByVal index As Integer) As Rectangle
        Dim selectedRow, selectedColumn As Integer
        selectedRow = CInt(System.Math.Floor((index - currentTopLeftItem) / colsPerPage))
        selectedColumn = (index - currentTopLeftItem) Mod colsPerPage
        Dim itemRect As New Rectangle((x) + (selectedColumn * ((2 * x) + w)), _
                                    (1 * y) + (selectedRow * ((2 * y) + h)), _
                                    w + (2 * x), h + (y * 2))
        Return itemRect
    End Function

    Private Sub m_PicBox_Resize(ByVal sender As Object, ByVal e As System.EventArgs) Handles m_PicBox.Resize
        CalculateRowsAndColumns()
        m_PicBox.Invalidate()
    End Sub

    Public Sub New(ByVal pPicBox As PictureBox)
        m_PicBox = pPicBox
        w = defaImageWidth
        h = defaImageHeight
        x = defaHorizontalSpacing
        y = defaVerticalSpacing

        CalculateRowsAndColumns()
        SelectedIndex = 0 '-1
    End Sub

    Private Function HitTest(ByVal loc As Point) As Integer
        Dim i As Integer
        Dim found As Boolean = False
        i = 0
        Do While i < List.Count And Not found
            If GetItemRect(i).Contains(loc) Then
                found = True
            Else
                i += 1
            End If
        Loop
        If found Then
            Return i
        Else
            Return -1
        End If
    End Function

    Protected Sub RefreshItem(ByVal index As Integer)
        Debug.WriteLine(String.Format("Refreshing {0}", index))
        Dim itemRect As Rectangle = GetItemRect(index)
        itemRect.Inflate(5, 5)
        m_PicBox.Invalidate(itemRect)
    End Sub

    Private Sub m_PicBox_MouseUp(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles m_PicBox.MouseUp
        'If e.Button = MouseButtons.Left Then
        'End If
        Dim mouseLoc As Point = New Point(e.X, e.Y) ' m_PicBox.PointToClient(m_PicBox.MousePosition)   'Me.PointToClient(Me.MousePosition())
        Dim itemHit As Integer = HitTest(mouseLoc)
        If itemHit <> -1 Then
            SelectedIndex = itemHit
        Else
            SelectedIndex = -1
        End If
        'm_PicBox.Focus()
        If e.Button = Windows.Forms.MouseButtons.Right Then
            If SelectedIndex > -1 Then
                fCacKyHieu.cmnuKyHieu.Show(fCacKyHieu.picThumbNails, New Point(e.X, e.Y))
            Else
                fCacKyHieu.cmnuThumbNail.Show(fCacKyHieu.picThumbNails, New Point(e.X, e.Y))
            End If
        ElseIf e.Button = Windows.Forms.MouseButtons.Left Then
            If SelectedIndex > -1 Then
                fMain.myBando.ChuanBiNhanKH()
            Else
                fMain.myBando.NhanKHXong()
            End If
        End If

        m_PicBox.Invalidate()

    End Sub

End Class
